src/util/datablock/datablock.c

```diff
@@ -72,93 +73,132 @@ void _DataBlock_AddBlocks(DataBlock *dataBlock, size_t blockCount) {
+// Make sure datablock can accommodate at least k items.
+void DataBlock_Accommodate(DataBlock *dataBlock, int64_t k) {
+    // Compute number of free slots.
+    int64_t freeSlotsCount = dataBlock->itemCap - dataBlock->itemCount;
+    int64_t additionalItems = k - freeSlotsCount;
+
+    if(additionalItems > 0) {
+        int64_t additionalBlocks = ITEM_COUNT_TO_BLOCK_COUNT(additionalItems);
+        _DataBlock_AddBlocks(dataBlock, additionalBlocks);
+    }
 }
 
-void DataBlock_SetItem(DataBlock *dataBlock, void *item, size_t idx) {
-    assert(dataBlock && idx <= dataBlock->itemCount);
+void *DataBlock_GetItem(const DataBlock *dataBlock, size_t idx) {
+    assert(dataBlock && idx >= 0);
 
-    // Get block.
-    size_t blockIdx = ITEM_INDEX_TO_BLOCK_INDEX(idx);
-    Block *block = dataBlock->blocks[blockIdx];
+    if(_DataBlock_IndexOutOfBounds(dataBlock, idx)) return NULL;
 
-    // Add item to block.
+    Block *block = GET_ITEM_BLOCK(dataBlock, idx);
     idx = ITEM_POSITION_WITHIN_BLOCK(idx);
-    size_t offset = idx * block->itemSize;
-    memcpy(block->data + offset , item, dataBlock->itemSize);
+    
+    char *item = block->data + (idx * block->itemSize);
+
+    // Incase item is marked as deleted, return NULL.
+    if(_DataBlock_IsItemDeleted(dataBlock, item)) return NULL;
+
+    return item;
 }
 
-void DataBlock_AddItems(DataBlock *dataBlock, size_t itemCount, DataBlockIterator **it) {
-    assert(dataBlock && itemCount > 0);
-    
+void* DataBlock_AllocateItem(DataBlock *dataBlock, u_int64_t *idx) {
     // Make sure we've got room for items.
-    if(dataBlock->itemCount + itemCount >= dataBlock->itemCap) {
+    if(dataBlock->itemCount >= dataBlock->itemCap) {
         // Allocate twice as much items then we currently hold.
-        size_t newCap = (dataBlock->itemCount + itemCount) * 2;
+        size_t newCap = dataBlock->itemCount * 2;
         size_t requiredAdditionalBlocks = ITEM_COUNT_TO_BLOCK_COUNT(newCap) - dataBlock->blockCount;
        _DataBlock_AddBlocks(dataBlock, requiredAdditionalBlocks);
     }
 
-    if(it) {
-        int step = 1;
-        Block *block = ACTIVE_BLOCK(dataBlock);
-        *it = DataBlockIterator_New(block, dataBlock->itemCount,
-                                    dataBlock->itemCount + itemCount, step);
+    // Get index into which to store item,
+    // prefer reusing free indicies.
+    u_int64_t pos = dataBlock->itemCount;
+    if(array_len(dataBlock->deletedIdx) > 0) {
+        pos = array_pop(dataBlock->deletedIdx);
     }
+    dataBlock->itemCount++;
 
-    dataBlock->itemCount += itemCount;
-}
+    if(idx) *idx = pos;
 
-Block *DataBlock_GetItemBlock(const DataBlock *dataBlock, size_t itemIdx) {
-    return GET_ITEM_BLOCK(dataBlock, itemIdx);
+    Block *block = GET_ITEM_BLOCK(dataBlock, pos);
+    pos = ITEM_POSITION_WITHIN_BLOCK(pos);
+    
+    char *item = block->data + (pos * block->itemSize);
+    _DataBlock_MarkItemAsUndelete(dataBlock, item);
```

Comment:
I don't see why this is necessary. If an item is undeleted and subsequently not modified at all by the caller, then accessing it after this would definitely cause a crash.

Reply:
I was thinking of a situation where we reuse a deleted item:
DataBlock_DeleteItem(dataBlock, 0);
void *item = DataBlock_AllocateItem(dataBlock, &idx);
not modifying item in anyway but then iterating over the datablock, in which case I don't want the iterator to think item at position 0 is deleted.